src/main/java/elucent/rootsclassic/block/altar/AltarBlockEntity.java

```diff
@@ -23,22 +25,30 @@
 import net.minecraft.world.level.block.entity.BlockEntityType;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.BlockHitResult;
+import net.minecraft.world.phys.Vec3;
 import net.minecraftforge.common.util.LazyOptional;
 import net.minecraftforge.items.IItemHandler;
 import net.minecraftforge.items.ItemHandlerHelper;
 import net.minecraftforge.items.ItemStackHandler;
 
 import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
 import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 
 public class AltarBlockEntity extends BEBase {
 	private static final int RECIPE_PROGRESS_TIME = 200;
-	private ArrayList<ItemStack> incenses = new ArrayList<>();
+	private List<ItemStack> incenses = Collections.emptyList();
 	private int ticker = 0;
 	private int progress = 0;
-	private ResourceLocation ritualName = null;
-	private RitualBase ritualCurrent = null;
-	//	private ItemStack resultItem = ItemStack.EMPTY; TODO: Unused
+
+	@Nullable
+	private RitualRecipe<?> ritualCurrent = null;
+	private int clientRitualLevel;
+	private BlockPos clientRitualColor = BlockPos.ZERO;
+	private BlockPos clientRitualSecondaryColor = BlockPos.ZERO;
```

Comment:
Any reason for storing the colors as Block positions? Could've changed it out entirely for a hex integer

Reply:
The colors where previously defined as integers & the altar tile used them as Vec3, but since the were doubles and there are handy built-in methods to encode/decode BlockPos to buffers I just used them.
But using a hex value in the JSON and a Java Color object would be better I think

Reply:
Or just simply using a hex integer